### DISCUSSION TITLE
Peer dependencies

### DISCUSSION THREAD

**Discussion by steida** [denoland/deno#15950] [Q&A]
There are a few old closed issues I read and to be honest, I still do not understand what is Deno official opinion on peer dependencies. While I understand it's not a problem on the server to bundle several different versions on the same dependency, for client-side development, it's a severe issue. Peer dependencies are not even mentioned in the documentation. What does it mean? Thank you.

**Comment by steida** [comment]
OK, so the question was moved to discussions 12 hours ago without an official response. Message received.

**Reply by bogdanbiv** [reply]
Note to self: aggressiveness always ensures a quick response, 10/10

**Reply by littledivy** [reply]
Going to drop the link to this discussion https://github.com/denoland/deno/issues/10356 where @kitsonk links many other issues related to the topic. 

> moved to discussions 12 hours ago

someone was kind enough to do house keeping on Sunday :) (People usually don't work on Sundays BTW)

**Reply by kitsonk** [reply]
I moved it because it was not a directly actionable issue, and more a discussion but at @littledivy I gave an official response a while back:

> The Deno CLI does not care what semantics a placed on the import specifier. That is up to the registry. There are several registries that already provide packages that way, like unpkg, snowpack, esm.sh, etc.
> 
> What you are suggesting is that you might want deno.land/x/ to support semver. That isn't an issue for the Deno CLI.
> 
> All that being said, this topic has been exhaustively been discussed before, so closing as duplicate of https://github.com/denoland/deno/issues/47, https://github.com/denoland/deno/issues/195, https://github.com/denoland/deno/issues/288, https://github.com/denoland/deno/issues/4574, https://github.com/denoland/deno/issues/6030 and others.

I'm sorry that you are finding this confusing.

**Reply by steida** [reply]
https://en.wikipedia.org/wiki/Dependency_hell

**Reply by Soremwar** [reply]
How interesting that you link that @steida , because by definition: "The dependency issue arises when several packages have dependencies on the same shared packages or libraries, but they depend on different and incompatible versions of the shared packages"

Which in turn would mean that peer dependencies can't solve this issue. That said Deno does provide a way to override subdependencies, it's called an import map

**Reply by KnorpelSenf** [reply]
There's no need to be sarcastic here. There's no silver bullet in dependency management, that's obvious, but import maps simply aren't satisfying. They require configuration for the library consumer. This is unacceptable. It has to be possible to copy-paste-run the code from my library's README. Explaining the issue of peer dependencies there and forcing users to create a JSON configuration file because of some internals is far from ideal.

I've said it in #10356 already, I'll repeat it here: Node can do it. (Yes, that means Bun can do it, too.) I wish we could come up with a solution for this for Deno. So far, this seems to be a pending task.

Our current solution for our ~30 plugins that support Deno is to rely on https://lib.deno.dev. This will redirect to the latest version. That way, each plugin automatically imports the latest version of the core library. That aligns perfectly with what users import in an ideal world. In practice, this breaks every once in a while, but people can fix it easily using `--reload`. This is becoming the Deno equivalent of `rm -rf node_modules && npm i`, which clearly isn't nice, but it's the least bad solution we have. It certainly works better for us than import maps do.

Perhaps someone can come up with something better, I'd be glad to hear about it.

**Reply by Soremwar** [reply]
I'm not trying to be sarcastic, but I'm sorry if I came out as too blunt

What I'm trying to say is that in this ecosystem where peer dependencies aren't widely available and every library mostly pings to a specific version we kinda sidestepped the dependency hell problem. Yes we will find some level of library duplication the more we dig into our dependency tree but at no point we'll find ourselves with incompatible versions of a library as we are bound to do in another ecosystem such as Node/NPM

I understand it has its drawbacks, but the more I use it the more convinced I am that if we delegate dedupping subdependencies to us (the end users) we'll find ourselves in less dependency hell situations. I understand that this means that the end user will always have to do dedupping manually, but in my opinion it's a better compromise than a project that can't be updated 1 year in the future out of fear of dependencies incompatibility

**Reply by KnorpelSenf** [reply]
That sounds great in theory and I actually believed the same thing in the beginning. In practice, we have some non-trivial type errors that result from users composing various elements from the plugins and the core library. They're incompatible (on the type level) if they stem from different minor versions.

We're already doing a fair bit of black TypeScript magic to avoid most of this, and sometimes we're able to cast types internally to make sure the errors go away, but this only gets us so far. Until then, people will have to `--reload` to do the manual deduping you're talking about.

In a way, peer deps could be seen as DI on the module level. I think it could be possible to move this to the code level via dynamic imports. I'm honestly not sure if that's any better due to the drawbacks of dynamic imports, what do you think?

**Reply by Soremwar** [reply]
> That sounds great in theory and I actually believed the same thing in the beginning. In practice, we have some non-trivial type errors that result from users composing various elements from the plugins and the core library. They're incompatible (on the type level) if they stem from different minor versions

I really don't understand what you mean by this. Typechecking will not be affected by importing different versions of the same library

> People will have to --reload to do the manual deduping you're talking about

`--reload` will do nothing for a pinned version

I'm kinda confused by what you describe here, could you elaborate @KnorpelSenf ?

**Reply by KnorpelSenf** [reply]
> > That sounds great in theory and I actually believed the same thing in the beginning. In practice, we have some non-trivial type errors that result from users composing various elements from the plugins and the core library. They're incompatible (on the type level) if they stem from different minor versions
> 
> I really don't understand what you mean by this. Typechecking will not be affected by importing different versions of the same library

Why not? In our case, users import a value (some deeply nested object which allows users to perform API calls) from one version, and a function consuming that object from another version, but if the API gets extended, a version mismatch will lead to compilation errors, even though no runtime code will fail.

> > People will have to --reload to do the manual deduping you're talking about
> 
> `--reload` will do nothing for a pinned version

As mentioned before, we use `lib.deno.dev` in order to update versions automatically through redirects, rather than pinning versions.

An alternative to our solution would be up ship dozens of patch updates for every release. Every plugin would have to be updated and the only change would be an updated dependency on the core lib.

**Reply by KnorpelSenf** [reply]
> In a way, peer deps could be seen as DI on the module level. I think it could be possible to move this to the code level via dynamic imports. I'm honestly not sure if that's any better due to the drawbacks of dynamic imports, what do you think?

By the way, this will not work well because Deno Deploy does not support dynamic imports—but we obviously want to be able to use the library there. So this is not a solution, either.

**Reply by steida** [reply]
> Note to self: aggressiveness always ensures a quick response, 10/10

Congrats https://deno.com/blog/package-json-support?utm_source=denonews&utm_medium=email#lingering-issues

**Reply by steida** [reply]
> I moved it because it was not a directly actionable issue, and more a discussion but at @littledivy I gave an official response a while back:
> 
> 
> 
> > The Deno CLI does not care what semantics a placed on the import specifier. That is up to the registry. There are several registries that already provide packages that way, like unpkg, snowpack, esm.sh, etc.
> 
> > 
> 
> > What you are suggesting is that you might want deno.land/x/ to support semver. That isn't an issue for the Deno CLI.
> 
> > 
> 
> > All that being said, this topic has been exhaustively been discussed before, so closing as duplicate of https://github.com/denoland/deno/issues/47, https://github.com/denoland/deno/issues/195, https://github.com/denoland/deno/issues/288, https://github.com/denoland/deno/issues/4574, https://github.com/denoland/deno/issues/6030 and others.
> 
> 
> 
> I'm sorry that you are finding this confusing.

Congrats to release package.json

**Reply by KnorpelSenf** [reply]
Supporting `package.json` makes a lot of sense if the goal is to support Node projects out of the box. In contrast, supporting bare specifiers and ambiguous imports and inventing new protocols such as `deno:` and `github:` is ridiculous. See https://github.com/denoland/deno/issues/17495#issuecomment-1480414548, I don't want to repeat all the arguments here.
